File: frontend/lib/model/instance/spaninstance/model/SpanInstance.ts
import Usage from "../../../phitagdata/usage/model/Usage";
import { IInstance, IInstanceConstructor } from "../../model/IInstance";
import SpanInstanceDto from "../dto/SentimentInstanceDto";
import SpanInstanceId from "./SpanInstanceId";

export default class SpanInstance implements IInstance {

    readonly id: SpanInstanceId;

    private readonly usage: Usage;

    private readonly labelSet: Array<string>;
    private readonly nonLabel: string;

    constructor(id: SpanInstanceId, usage: Usage, labelSet: Array<string>, nonLabel: string) {
        this.id = id;
        this.usage = usage;
        this.labelSet = labelSet;
        this.nonLabel = nonLabel;
    }

    public getId(): SpanInstanceId {
        return this.id;
    }

    public getUsage(): Usage {
        return this.usage;
    }

    public getLabelSet(): Array<string> {
        return this.labelSet;
    }

    public getNonLabel(): string {
        return this.nonLabel;
    }

    public getLabelAndNonLabel(): Array<string> {
        return this.labelSet.concat(this.nonLabel);
    }

    public static fromDto(dto: SpanInstanceDto): SpanInstance {
        return new SpanInstance(
            SpanInstanceId.fromDto(dto.id),
            Usage.fromDto(dto.usage),
            dto.labelSet,
            dto.nonLabel
        );
    }
    
}

export class SpanInstanceConstructor implements IInstanceConstructor {
    fromDto(dto: SpanInstanceDto): SpanInstance {
        return SpanInstance.fromDto(dto);
    }
}

File: backend/src/main/java/de/garrafao/phitag/domain/instance/spaninstance/SpanInstance.java
package de.garrafao.phitag.domain.instance.spaninstance;

import de.garrafao.phitag.domain.instance.IInstance;
import de.garrafao.phitag.domain.phase.Phase;
import de.garrafao.phitag.domain.phitagdata.usage.Usage;
import lombok.Getter;
import org.apache.commons.lang3.Validate;

import javax.persistence.*;
import java.util.Arrays;
import java.util.List;

@Entity
@Table(name = "phitagspaninstance")
@Getter
public class SpanInstance implements IInstance {

    @EmbeddedId
    private SpanInstanceId id;

    @MapsId("phaseid")
    @ManyToOne
    @JoinColumns({
            @JoinColumn(name = "phasename", referencedColumnName = "name"),
            @JoinColumn(name = "projectname", referencedColumnName = "projectname"),
            @JoinColumn(name = "ownername", referencedColumnName = "ownername")
    })
    private Phase phase;

    @ManyToOne
    @JoinColumns({
            @JoinColumn(name = "phitagusage_dataid", referencedColumnName = "dataid"),
            @JoinColumn(name = "phitagusage_projectname", referencedColumnName = "projectname"),
            @JoinColumn(name = "phitagusage_ownername", referencedColumnName = "ownername")
    })
    private Usage usage;

    @Column(name = "label_set", nullable = false)
    private String labelSet;

    @Column(name = "non_label", nullable = false)
    private String nonLabel;

    public SpanInstance() {
    }

    public SpanInstance(
            final String instanceId, final Phase phase,
            final Usage usage, final String labelSet, final String nonLabel) {
        Validate.notNull(instanceId);
        Validate.notNull(phase);
        Validate.notNull(usage);

        this.id = new SpanInstanceId(instanceId, phase.getId());
        this.phase = phase;
        this.usage = usage;
        this.labelSet = labelSet;
        this.nonLabel = nonLabel;
    }

    public List<String> getLabelSet() {
        return Arrays.asList(labelSet.split(","));
    }

    @Override
    public String toString() {
        return String.format("InstanceData [id=%s]", id);
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) {
            return true;
        }

        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }

        final SpanInstance other = (SpanInstance) obj;
        return this.id.equals(other.id);
    }

    @Override
    public int hashCode() {
        return this.id.hashCode();
    }

}
